google-cloud-firestore/src/test/java/com/google/cloud/firestore/LocalFirestoreHelper.java

```diff
@@ -889,4 +928,72 @@ public static String autoId() {
   public static Map<String, Object> fromSingleQuotedString(String json) {
     return fromJsonString(json.replace("'", "\""));
   }
+
+  static class RequestResponseMap
+      extends LinkedHashMap<GeneratedMessageV3, ApiFuture<? extends GeneratedMessageV3>> {
+    int activeRequestCounter = 0;
+    SettableApiFuture<Void> activeRequestComplete = SettableApiFuture.create();
+    CountDownLatch latch = new CountDownLatch(1);
+
+    void markComplete() {
+      activeRequestComplete.set(null);
+    }
+
+    void awaitLatch() {
+      try {
+        latch.await();
```

Comment:
You should try to generalized this code if you keep it in this generalized helper. A good way to do this would be a BlockingQueue of requests. This could be `awaitRequest` and use `BlockingQueue.poll()` to wait for the next request. You would supply the requests in the stub handler (and call `BlockingQueue.offer()`).

Reply:
Does this modified implementation work? I'm still using a single latch and future, but it can process multiple serial requests.

Reply:
If I follow this code correctly, you cannot wait for `n` requests or buffer `n` requests. You have to send, wait, send, wait, send and so on. If I send 2 requests, I should be able to invoke `awaitRequest()` twice without blocking.

Reply:
I changed it from CountDownLatch to Semaphore, and I think you can buffer multiple requests now. Buffering `n` requests releases `n` permits that `n` `awaitRequest()` calls can instantly consume without blocking.